enhancements/monitoring/replace-prometheus-adapter-with-metrics-server.md

```diff
@@ -0,0 +1,257 @@
+
+#### Story 2
+
+* As a cluster admin, I seek a lightweight and efficient solution with improved accuracy for monitoring resource metrics within my cluster.
+
+#### Story 3
+
+* As a cluster admin, I want a simplified monitoring architecture in OpenShift, so that I can reduce the complexity and maintenance overhead of the monitoring system.
+
+### Goals
+
+* Replace Prometheus Adapter with Metrics Server while maintaining feature parity with the existing stack 
+* Ensure no performance regression occurs when replacing Prometheus Adapter with Metrics Server
+* Improved scale up time for autoscaling
+
+### Non-Goals
+
+* Allow optional cluster monitoring stack
+* Moving `openshift-monitoring/cluster-monitoring-config` config-map to a Custom Resource
+
+## Proposal
+
+The proposal is to replace the Prometheus Adapter with the Metrics Server for monitoring resource metrics in OpenShift.
+
+### Workflow Description
+
+The workflow for using the Metrics Server in OpenShift for monitoring resource metrics would involve the following steps:
+
+1. User sets `enabled: true` in the ConfigMap `openshift-monitoring/cluster-monitoring-config` under `metricsServer` section
+
+    ```yaml
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: cluster-monitoring-config
+      namespace: openshift-monitoring
+    data:
+      config.yaml: |
+        metricsServer:
+          enabled: true
+    ```
+
+With this configuration, the cluster monitoring operator should:
+  1. Deploy the metrics-server resources in the `openshift-monitoring` namespace
+  1. Verify that the `metrics-server` service is functional by querying the `/readyz` endpoint
+  1. Wait for `5 x metrics-server's scrape interval` to have enough data populated in `metrics-server` in-memory
+  1. Update `APIService` resource to point to `metrics-server` API instead of `prometheus-adapter` API
+  1. Delete `prometheus-adapter` resources
+
+#### Variation [optional]
+
+NA
+
+### API Extensions
+
+- We propose to add type `metricsServer` to the `openshift-montoring/cluster-monitoring-config` configmap to define 
+  metrics-server settings. 
+
+  ```go
+  type ClusterMonitoringConfiguration struct {
+    ...
+    // `MetricsServerConfig` defines settings for the MetricsServer component.
+    MetricsServerConfig *MetricsServerConfig `json:"metricsServer,omitempty"`
+  }
+  ```
+
+  When `enabled` is set `true`, the cluster monitoring operator configures the Metrics Server resources in `openshift-monitoring` namespace and tears down the Prometheus adapter resources once Metrics Server is fully functional.
+
+  ```go
+  // The `MetricsServerConfig` resource defines settings for the MetricsServer component.
+  type MetricsServerConfig struct {
+    // `Enabled` is a Boolean flag that enables or disables metrics-server.
+    Enabled *bool `json:"enabled,omitempty"`
+    // Defines the nodes on which the pods are scheduled.
+    NodeSelector map[string]string `json:"nodeSelector,omitempty"`
```

Comment:
I think NodeSelector is a good name here, but I just want to mention that a --node-selector option was recently added to metrics-server https://github.com/yangjunmyfm192085/metrics-server/blob/8ac883d7747e53a459e49ee30e3827089f593430/docs/command-line-flags.txt#L23 and it is doing something a bit different than this. It is not released yet, but if we ever enable downstream, it might become confusing, unless it is part of some kind of Kubelet configuration inside of the MetricsServerConfig.

Reply:
ok IIUC new field is to select nodes to scrape metrics. NodeSelector was used for all other components, may be we should use another name here considering same upstream flag will be upcoming for metrics-server

Reply:
Thanks for the heads-up! I'd stick with `nodeSelector` here to be consistent with the other monitoring components. When CMO wants to support the new `--node-selector` argument, we'll need to make it explicit which setting applies to what (either by naming or documentation or both).

But we may not look for a 1:1 API mapping between CMO and metrics-server. If the main goal is to skip nodes, we might as well use a predefined label (e.g. `openshift.io/node-metrics: false`) and users would put it on nodes that shouldn't be queried by metrics-server.

Reply:
Yes the new flag is to the select the nodes to scrape metrics on.

Yeah I also think that sticking with `nodeSelector` here and potentially diverging from upstream later is the best option from a UX perspective.